src/simplnx/Utilities/ThreadedAlgorithm.hpp

```diff
@@ -0,0 +1,46 @@
+#pragma once
+
+#include "simplnx/Filter/IFilter.hpp"
+#include "simplnx/simplnx_export.hpp"
+
+#include <chrono>
+#include <cstdint>
+
+namespace nx::core
+{
+
+class SIMPLNX_EXPORT ThreadedAlgorithm
+{
+
+public:
+  ThreadedAlgorithm(const IFilter::MessageHandler& mesgHandler, const std::atomic_bool& shouldCancel);
+  virtual ~ThreadedAlgorithm() noexcept;
+
+  ThreadedAlgorithm(const ThreadedAlgorithm&) = delete;
+  ThreadedAlgorithm(ThreadedAlgorithm&&) noexcept = delete;
+  ThreadedAlgorithm& operator=(const ThreadedAlgorithm&) = delete;
+  ThreadedAlgorithm& operator=(ThreadedAlgorithm&&) noexcept = delete;
+
+  const std::atomic_bool& getCancel() const;
+
+  void sendThreadSafeProgressMessage(size_t counter);
```

Comment:
```suggestion
/**
   * @brief This function accepts a number to increment the counter by and handles all of the message 
   * assembly and send off
   * [!!Utilizes Mutex Locks!!] be wary calling this excessively in parallel as it can slow the program down
   * @param counter the number of elements the function has processed since last update progress call
   */
  void sendThreadSafeProgressMessage(size_t counter);
```